tests/modeladmin/tests.py

```diff
@@ -175,13 +189,72 @@ class WaiterAdmin(ModelAdmin):
             ]
 
         ma = WaiterAdmin(Waiter, self.site)
-        self.assertIs(ma.lookup_allowed("restaurant__place__country", "1"), True)
+        mock_request = MockRequest()
+        self.assertIs(
+            ma.lookup_allowed("restaurant__place__country", "1", request=mock_request),
+            True,
+        )
+        self.assertIs(
+            ma.lookup_allowed(
+                "restaurant__place__country__id__exact", "1", request=mock_request
+            ),
+            True,
+        )
+        self.assertIs(
+            ma.lookup_allowed(
+                "restaurant__place__country__name", "test_value", request=mock_request
+            ),
+            True,
+        )
+
+    def test_lookup_allowed_considers_dynamic_list_filter(self):
+        class ConcertAdmin(ModelAdmin):
+            list_filter = ["main_band__sign_date"]
+
+            def get_list_filter(self, request):
+                return self.list_filter + ["main_band__name"]
+
+        model_admin = ConcertAdmin(Concert, self.site)
+        request_band_name_filter = RequestFactory().get(
+            "/", {"main_band__name": "test"}
+        )
         self.assertIs(
-            ma.lookup_allowed("restaurant__place__country__id__exact", "1"), True
+            model_admin.lookup_allowed(
+                "main_band__sign_date", "?", request=request_band_name_filter
+            ),
+            True,
         )
         self.assertIs(
-            ma.lookup_allowed("restaurant__place__country__name", "test_value"), True
+            model_admin.lookup_allowed(
+                "main_band__name", "?", request=request_band_name_filter
+            ),
+            True,
+        )
+
+    def test_lookup_allowed_without_request_deprecation(self):
+        class ConcertAdmin(ModelAdmin):
+            list_filter = ["main_band__sign_date"]
+
+            def get_list_filter(self, request):
+                return self.list_filter + ["main_band__name"]
+
+            def lookup_allowed(self, lookup, value):
+                return True
+
+        model_admin = ConcertAdmin(Concert, self.site)
+        msg = (
+            "Add 'request' to the signature of ConcertAdmin.lookup_allowed(). Support "
+            "for ModelAdmin subclasses that do not accept this argument is deprecated."
+        )
+        request_band_name_filter = RequestFactory().get(
+            "/", {"main_band__name": "test"}
+        )
+        request_band_name_filter.user = User.objects.create_superuser(
+            username="bob", email="bob@test.com", password="test"
         )
+        with self.assertRaisesMessage(RemovedInDjango60Warning, msg):
+            changelist = model_admin.get_changelist_instance(request_band_name_filter)
+            changelist.get_filters(request_band_name_filter)
```

Comment:
This line is unreachable.

We could use `assertWarnsMessage()` instead of `assertRaisesMessage()` and check if it still works, e.g.

```python
with self.assertWarnsMessage(RemovedInDjango60Warning, msg):
    changelist = model_admin.get_changelist_instance(request_band_name_filter)
    filterspec = changelist.get_filters(request_band_name_filter)[0][0]
    self.assertEqual(filterspec.title, "sign date")
    filterspec = changelist.get_filters(request_band_name_filter)[0][1]
    self.assertEqual(filterspec.title, "name")
    self.assertSequenceEqual(filterspec.lookup_choices, [self.band.name])
```